### DISCUSSION TITLE
Fix Payjoin constraints limiting widespread adoption

### DISCUSSION THREAD

**Discussion by DanGould** [btcpayserver/btcpayserver#4960] [Ideas / Feature Requests]
The Human Rights Foundation has 2 separate donation portals. [One for bitcoin and lightning](https://hrf.org/btc), and [one for payjoin](https://hrf.org/payjoin). One of the great benefits of payjoin is that it fits in unified URIs & QR codes. Separating them defeats the purpose of a unified QR and loses many opportunities to payjoin by default without user knowhow.

They two stores are separated because HRF wants to secure as much as they can with cold storage, and BTCPay Payjoin requires hot wallet only. I see two ways to address this problem.

## Fix 1: Payjoin Output Substitution to Cold Storage

Payjoin output substitution lets the receiver direct inbound funds to any address during receipt without including them in the URI or QR. This works because the sender requests a Payjoin PSBT after offering their input and transfer where the receiver can include new addresses.

If the Payjoin module could request both cold storage addresses and hot wallet addresses, it could consolidate funds to cold storage based on application logic. When a payjoin is incoming, if the incoming amount exceeds a threshold, or if the hot wallet balance does, the Payjoin PSBT could accept inbound sats and send funds to cold storage at once (and keep sats in the hot wallet for future payjoin, too). This solves the custody security problem and allows unified QR to offer payjoin on the same store that typically directs funds to cold storage.

This setup would still require a store owner to create and back up a hot wallet, but banners could guide owners through this setup step to be done by default.

## Fix 2: Use lightning wallet as payjoin (or store-wide) hot wallet

Other than engineering complexity, I see no security reason not to use a lightning capable wallet for both store hot wallet and Payjoin. Combining these wallets would eliminate the need for two stores by HRF and allow BTCPayServer to open lightning channels directly from inbound funds as Payjoin outputs.

However, I acknowledge that internal architecture of BTCPayServer may make this a difficult engineering problem in practice, so I suggest Fix #1 to address the immediate problem and to explore Fix 2 as aspiration to keep in mind as BTCPay develops.

[Nolooking](https://github.com/chaincase-app/nolooking) demonstrates using a payjoin with LND's hot wallet and RPC. LND maintains at least one hot UTXO as anchor which is available to contribute to Payjoin. Nolooking is written to support lightning rpc as an abstraction, so CLN cold be supported in this model too, it just haven't been wired together yet.

---

Note: I've edited the original post for clarity. To make better sense of the first comment one may wish to view the post before edits.

**Comment by Kukks** [comment]
Can you link these portals so we see what he means? I'm guessing he set up two stores on btcpay where one is with a hot wallet and payjoin, and the other is with a cold wallet? And both with the same Ln node?

I'm not sure what the request here is. Is it to  make lnds hot wallet used in btcpay? We have to build features agnostic of the lightning nodes connected to a store so nothing should be lnd specific.

**Reply by DanGould** [reply]
Updated OP with links to 2 separate stores. Seems like the payjoin store has no lightning option. Afaict they want 2 separate stores so they don't need a hot wallet to receive main donations. I propose one solution is to combine the hot wallet of the lightning node with the one that makes payjoin utxos available, as demonstrated with [Nolooking](https://github.com/chaincase-app/nolooking) on LND (though it could be adapted for another implementation too)

**Reply by dennisreimann** [reply]
I recently did a bit of research on this and the gist is: LND and CLN offer ways to export the wallet‘s xpub – which can then be imported to BTCPay Server as a watch-only onchain wallet. However, to my knowledge there is no way to share the hot wallet of BTCPay with the LN node or the other way around. 

Also see [LND key import](https://docs.lightning.engineering/lightning-network-tools/lnd/key_import) and [lightning-hsmtool](https://lightning.readthedocs.io/lightning-hsmtool.8.html).

**Reply by DanGould** [reply]
My idea would be to use the PSBT rpc interface to do operations and make use of their utxo DB rather than import privkeys. That's how nolooking works with LND. Seems like it could work the same for CLN too.

**Reply by Kukks** [reply]
This would not be feasible based on the amount of effort required IMO. As their LN node hot wallet is not actively receiving donations, their only source of getting new UTXOs is
* they manually send coins to open channels and leave some change
* a channel is closed from either party

This means that there would not be many utxos to payjoin with
I think the resolution here for them would be
* enable lightning on both stores regardless of cold/hot onchainw wallet as it is supported this way.
* Modify the Prism plugin to work on-chain
* Modify the prism plugin to allow a source to be a catch-all on the actual balance of the store's wallet
*  Configure an automated payout processor to transfer funds when the balance threshold is reached
*  Work on supporting payment batching within payjoin and then replace the automated payment processor with it

**Reply by Kukks** [reply]
> I recently did a bit of research on this and the gist is: LND and CLN offer ways to export the wallet‘s xpub – which can then be imported to BTCPay Server as a watch-only onchain wallet. However, to my knowledge there is no way to share the hot wallet of BTCPay with the LN node or the other way around.
> 
> Also see [LND key import](https://docs.lightning.engineering/lightning-network-tools/lnd/key_import) and [lightning-hsmtool](https://lightning.readthedocs.io/lightning-hsmtool.8.html).

Doing this potentially creates gap limit issues on the lightning node tracking and also introduces the risk of address reuse (an old invoice address is used by the node to request coins to be sent to it, forcing the invoice to register a payment when it was meant for an entirely different party)
